modules/build.gradle.kts

```diff
@@ -0,0 +1,25 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+plugins {
+    `java-platform`
+}
+
+javaPlatform {
+    allowDependencies()
+}
+
+dependencies {
+    // This platform depends on each of its subprojects.
+    subprojects {
+        runtime(this)
```

Comment:
I figure this builds the runtime classpath so all the module jars are available to the engine. `this` can be used to refer to a whole project _level_ in Gradle, so to say?

This reminds me of that old idea to make the PC facade have a compile dependency on all the modules, to have them for sure get built when you run the game. Although at the cost of not being able to launch without every single module working, and maybe at a risk to get weird dependencies written into the binaries published to Artifactory.

Reply:
`this` is given by the closest enclosing scope; it's the project that subprojects is iterating through. (the hints added by IntelliJ make this clearer when reading it in the IDE)

> make the PC facade have a compile dependency on all the modules

Yes, that's exactly what is happening when the facade depends on the platform.

> at the cost of not being able to launch without every single module working

I think we already sacrificed that ability a while ago?

Test plan:

* `groovyw recurse JoshariasSurvival`
* add some compilation error to EdibileFlora's FillingModifierSystem
* `gradlew game --continue`
  * `:modules:EdibleFlora:compileJava` fails

I think because the `game` task depends on the `moduleClasses` task.

I haven't yet come across a gradle way to always _attempt_ to rebuild dependency subprojects if their sources changed but also soft-fail and try to run subsequent steps despite that. You can `--continue`, but I think that only makes it keep going on independent parts of the dependency tree, not excuse a missing dependency.

Reply:
> get weird dependencies written into the binaries published to Artifactory.

yes, if you publish a jar of the PC facade while you have subprojects present in `/modules/`, that would include things we probably don't want included.

Is that a thing that happens? Does anything publish facade that's not in a clean workspace made explicitly for that purpose?

If so, then yeah, we'll have to add more separation so this local workspace state isn't confused with what we want to publish to the world. I hadn't imagined it was used that way.

Reply:
Gradle will always hit the issues, yep, IntelliJ has had a tendency in the past to skip errors if they're in something the run config doesn't care about. I'm not sure about that now since we're having some other weird issues there from time to time :-)

Reply:
> if you publish a jar of the PC facade ...

I didn't see this comment earlier while posting my own for some reason. No, that isn't done anywhere at the moment. And any sort of test-publishing we might do locally surely should target something like the Nanoware test repos in Artifactory.

It is good to know though, in case we get creative and weird in the future. One idea that floated around was to build modules in an actual full engine workspace, rather than the minified single-module workspace we run in Jenkins right now. Even that _in theory_ shouldn't hit it - since that should just publish the module, not the facade.

So - not worried about it. May be worth highlighting with a comment perhaps :-)